components/reply.js

```diff
@@ -55,61 +56,76 @@ export default forwardRef(function Reply ({
-      mutation upsertComment($text: String!, $parentId: ID!, $hash: String, $hmac: String) {
-        upsertComment(text: $text, parentId: $parentId, hash: $hash, hmac: $hmac) {
-          ...CommentFields
-          deleteScheduledAt
-          reminderScheduledAt
-          comments {
-            ...CommentsRecursive
+  const [upsertComment] = usePaidMutation(CREATE_COMMENT, {
+    update (cache, { data: { upsertComment: { result } } }) {
+      if (!result) return
+
+      cache.modify({
+        id: `Item:${parentId}`,
+        fields: {
+          comments (existingCommentRefs = []) {
+            const newCommentRef = cache.writeFragment({
+              data: result,
+              fragment: COMMENTS,
+              fragmentName: 'CommentsRecursive'
+            })
+            return [newCommentRef, ...existingCommentRefs]
           }
         }
-      }`, {
-      update (cache, { data: { upsertComment } }) {
+      })
+
+      const ancestors = item.path.split('.')
+
+      // update all ancestors
+      ancestors.forEach(id => {
         cache.modify({
-          id: `Item:${parentId}`,
+          id: `Item:${id}`,
           fields: {
-            comments (existingCommentRefs = []) {
-              const newCommentRef = cache.writeFragment({
-                data: upsertComment,
-                fragment: COMMENTS,
-                fragmentName: 'CommentsRecursive'
-              })
-              return [newCommentRef, ...existingCommentRefs]
+            ncomments (existingNComments = 0) {
+              return existingNComments + 1
             }
           }
         })
+      })
 
-        const ancestors = item.path.split('.')
+      // so that we don't see indicator for our own comments, we record this comments as the latest time
+      // but we also have record num comments, in case someone else commented when we did
+      const root = ancestors[0]
+      commentsViewedAfterComment(root, result.createdAt)
+    }
+  })
 
-        // update all ancestors
-        ancestors.forEach(id => {
-          cache.modify({
-            id: `Item:${id}`,
-            fields: {
-              ncomments (existingNComments = 0) {
-                return existingNComments + 1
-              }
+  const onSubmit = useCallback(async (variables, { resetForm }) => {
+    await upsertComment({
+      variables: { parentId, ...variables },
+      onCompleted: (data) => {
+        resetForm({ text: '' })
+        setReply(replyOpen || false)
+        toastUpsertSuccessMessages(toaster, data, 'upsertComment', false, variables.text)
+      },
+      onPaid: (cache, { data: { upsertComment: { result: { id } } } }) => {
+        // set the invoiceActionState to PAID
+        cache.modify({
+          id: `Item:${id}`,
+          fields: {
+            invoiceActionState () {
+              return 'PAID'
             }
-          })
+          }
+        })
+      },
+      onPayError: (e, cache, { data: { upsertComment: { result: { id } } } }) => {
```

Comment:
This throws `Cannot read properties of null (reading 'id')` if I am closing the QR code invoice as anon because `result` is null in that case.

So I think you need to check `paymentMethod` first. The cache update only needs to run if it was optimistic.